_source/_posts/2021-12-16-cloud-native-java-microservices-with-istio.md

```diff
@@ -0,0 +1,580 @@
+}
+
+service Customer with serviceClass
+paginate Customer with pagination
+
+/**
+ * Entities for product microservice
+ */
+// Product sold by the Online store
+entity Product {
+    name String required
+    description String
+    price BigDecimal required min(0)
+    itemSize Size required
+    image ImageBlob
+}
+
+enum Size {
+    S, M, L, XL, XXL
+}
+
+entity ProductCategory {
+    name String required
+    description String
+}
+
+entity ProductOrder {
+    placedDate Instant required
+    status OrderStatus required
+    code String required
+    invoiceId Long
+    customer String required
+}
+
+enum OrderStatus {
+    COMPLETED, PENDING, CANCELLED
+}
+
+entity OrderItem {
+    quantity Integer required min(0)
+    totalPrice BigDecimal required min(0)
+    status OrderItemStatus required
+}
+
+enum OrderItemStatus {
+    AVAILABLE, OUT_OF_STOCK, BACK_ORDER
+}
+
+relationship ManyToOne {
+	OrderItem{product(name) required} to Product
+}
+
+relationship OneToMany {
+   ProductOrder{orderItem} to OrderItem{order(code) required} ,
+   ProductCategory{product} to Product{productCategory(name)}
+}
+
+service Product, ProductCategory, ProductOrder, OrderItem with serviceClass
+paginate Product, ProductOrder, OrderItem with pagination
+microservice Product, ProductOrder, ProductCategory, OrderItem with product
+
+/**
+ * Entities for Invoice microservice
+ */
+// Invoice for sales
+entity Invoice {
+    code String required
+    date Instant required
+    details String
+    status InvoiceStatus required
+    paymentMethod PaymentMethod required
+    paymentDate Instant required
+    paymentAmount BigDecimal required
+}
+
+enum InvoiceStatus {
+    PAID, ISSUED, CANCELLED
+}
+
+entity Shipment {
+    trackingCode String
+    date Instant required
+    details String
+}
+
+enum PaymentMethod {
+    CREDIT_CARD, CASH_ON_DELIVERY, PAYPAL
+}
+
+relationship OneToMany {
+    Invoice{shipment} to Shipment{invoice(code) required}
+}
+
+service Invoice, Shipment with serviceClass
+paginate Invoice, Shipment with pagination
+microservice Invoice, Shipment with invoice
+
+/**
+ * Entities for notification microservice
+ */
+entity Notification {
+    date Instant required
+    details String
+    sentDate Instant required
+    format NotificationType required
+    userId Long required
+    productId Long required
+}
+
+enum NotificationType {
+    EMAIL, SMS, PARCEL
+}
+
+microservice Notification with notification
+```
+
+We define entities for each service and mark the entities as microservice entities. We also define relationships between entities, enums and other options like pagination, service layer and so on. Please refer to JDL [Entities](https://www.jhipster.tech/jdl/entities-fields) and [relationships](https://www.jhipster.tech/jdl/relationships) documentation for more possibilities.
+
+Now, we are ready to run JHipster. Open a terminal window on the folder where you saved the JDL and run the below command.
```

Comment:
I think you should mention (and maybe even show) that you should create a directory for everything.

```
mkdir jhipster-istio
cd jhipster-istio
```